File: go/src/aoc-2021-go/day1/day1.go
package main

import (
	. "aoc-2021-go/utils"
	"bufio"
	"embed"
	"fmt"
	"strconv"
)

//go:embed input.txt
var fEmbed embed.FS

func main() {
	numbers := parseInput()

	part1(numbers)
	part2(numbers)
}

func part1(numbers []int) {
	var increases = 0

	pairs := Zip(numbers, numbers[1:])
	for _, pair := range pairs {
		if pair.A < pair.B {
			increases++
		}
	}

	fmt.Printf("Part1: There were %d increases\n", increases)
}

func part2(numbers []int) {
	// Insight: there is no need to actually calculate the sliding window of 3 values. When comparing (a, b, c) to (b, c, d),
	// we know that the difference between the two is just -a+d. Therefore, the second 3-window tuple is larger iff (d>a).
	// Therefore, we can use an identical approach as part one, just zipping with n+4 instead of n+1.

	var increases = 0

	for _, pair := range Zip(numbers, numbers[3:]) {
		if pair.A < pair.B {
			increases++
		}
	}

	fmt.Printf("Part2: There were %d increases\n", increases)
}

func parseInput() []int {
	var numbers []int

	f, err := fEmbed.Open("input.txt")
	Check(err)

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		var line = scanner.Text()
		var value, err = strconv.Atoi(line)
		Check(err)

		numbers = append(numbers, value)
	}
	return numbers
}

File: rust/src/day1.rs
pub mod day1 {
    pub fn day1() {
        let numbers = parse_input();

        part1(numbers.as_slice());
        part2(numbers.as_slice());
    }

    fn part1(numbers: &[i32]) {
        let increases = numbers.iter().zip(numbers.iter().skip(1)).filter(|(a, b)| a < b).count();
        println!("Part 1: There were {} increases", increases)
    }

    fn part2(numbers: &[i32]) {
        // Insight: there is no need to actually calculate the sliding window of 3 values. When comparing (a, b, c) to (b, c, d),
        // we know that the difference between the two is just -a+d. Therefore, the second 3-window tuple is larger iff (d>a).
        // Therefore, we can use an identical approach as part one, just zipping with n+3 instead of n+1.

        let increases = numbers.iter().zip(numbers.iter().skip(3)).filter(|(a, b)| a < b).count();
        println!("Part 2: There were {} increases", increases)
    }

    fn parse_input() -> Vec<i32> {
        let bytes = include_bytes!("../assets/day1.txt");
        let string = String::from_utf8_lossy(bytes);
        string.lines().map(|line| line.parse::<i32>().unwrap()).collect::<Vec<i32>>()
    }
}

#[cfg(test)]
mod tests {
    use crate::day1;

    #[test]
    fn run_it() {
        day1()
    }
}